### PR TITLE
DM-23049: Fix association of flag columns to forced_src catalogs

### PR DESCRIPTION
**Pull request by laurenam**

### CODE REVIEW

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 1]
```diff
@@ -251,22 +259,23 @@ def runDataRef(self, patchRefList, subdir="", cosmos=None):
             unforced = self.calibrateCatalogs(unforced, wcs=repoInfo.wcs)
             plotKwargs.update(dict(zpLabel=self.zpLabel))
             if haveForced:
-                # copy over some fields from unforced to forced catalog
-                forced = addColumnsToSchema(unforced, forced,
-                                            [col for col in list(self.config.columnsToCopy) +
-                                             list(self.config.analysis.flags) if
-                                             col not in forced.schema and col in unforced.schema and
-                                             not (repoInfo.hscRun and col == "slot_Centroid_flag")])
```
Comment (sr525) [inline python/lsst/pipe/analysis/coaddAnalysis.py:259]: I think this would be more legible and easier to understand if the list was defined prior to the function call.
 eg. colsTocopy = [col for col in ....], addColumnsToSchema(unforced, forced, colsToCopy).
Where is columnsToCopy defined? Was it not replaced by the new config option? I think that both of the config values are already lists and probably don't need casting to list.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 2]
```diff
@@ -251,22 +259,23 @@ def runDataRef(self, patchRefList, subdir="", cosmos=None):
...
-                forced = addColumnsToSchema(refBandCat, forced,
-                                            [col for col in refBandList if col not in forced.schema and
-                                             col in refBandCat.schema])
+                                       format(len(forced), len(refCat)))
+                refColList = [s for s in refCat.schema.getNames() if
+                              s.startswith(tuple(self.config.columnsToCopyFromRef))]
+                forced = addColumnsToSchema(refCat, forced,
+                                            [col for col in refColList if col not in forced.schema and
+                                             col in refCat.schema and not
+                                             (repoInfo.hscRun and col == "slot_Centroid_flag")])
```
Comment (sr525) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: Same comment as above.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 3]
```diff
@@ -251,22 +259,23 @@ def runDataRef(self, patchRefList, subdir="", cosmos=None):
...
+                measColList = [s for s in unforced.schema.getNames() if
+                               s.startswith(tuple(self.config.columnsToCopyFromMeas))]
+                forced = addColumnsToSchema(unforced, forced,
+                                            [col for col in measColList if col not in forced.schema and
+                                             col in unforced.schema and not
+                                             (repoInfo.hscRun and col == "slot_Centroid_flag")])
```
Comment (sr525) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: And here

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 4]
```diff
@@ -477,16 +491,29 @@ def readSrcMatches(self, dataRefList, dataset, hscRun=None, wcs=None, aliasDictL
             # Set some aliases for differing schema naming conventions
             if aliasDictList:
                 catalog = setAliasMaps(catalog, aliasDictList)
-            if dataset != "deepCoadd_meas" and any(ss not in catalog.schema
-                                                   for ss in self.config.columnsToCopy):
-                unforced = dataRef.get("deepCoadd_meas", immediate=True,
+            if dataset != (self.config.coaddName + "Coadd_meas" and
+                           any(ss not in catalog.schema for ss in list(self.config.columnsToCopyFromMeas) +
+                               list(self.config.columnsToCopyFromRef))):
```
Comment (sr525) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: I don't think these need to be cast as lists either and it might improve readability if the any(...) is on another line.
e.g. 

`colsNotInCatalogMask = any(col not in catalog.schema for col in self.config.columnsToCopyFromMeas + self.config.columnsToCopyFromRef)` 
`if dataset != (self.config.coaddName + "Coadd_meas" and colsNotInCatalogMask)`

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 5]
```diff
@@ -477,16 +491,29 @@ def readSrcMatches(self, dataRefList, dataset, hscRun=None, wcs=None, aliasDictL
...
-                # copy over some fields from unforced to forced catalog
+                if len(catalog) != len(refCat):
+                    raise RuntimeError(("Lengths of forced (N = {0:d}) and ref (N = {0:d}) cats don't match").
+                                       format(len(catalog), len(refCat)))
+                refColList = [s for s in refCat.schema.getNames() if
+                              s.startswith(tuple(self.config.columnsToCopyFromRef))]
+                catalog = addColumnsToSchema(refCat, catalog,
+                                             [col for col in refColList if col not in catalog.schema and
+                                             col in refCat.schema and not
+                                             (hscRun and col == "slot_Centroid_flag")])
```
Comment (sr525) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: This seems to be the same list as earlier and could be reused. It would also be cleaner to reuse refColList rather than redefining it.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 6]
```diff
@@ -95,13 +95,21 @@ class CoaddAnalysisConfig(Config):
...
-                                       "detect_isTractInner", "merge_peak_sky", "calib_psfUsed",
-                                       "calib_psfCandidate", ],
-                              doc="List of columns to copy from one source catalog to another.")
+    # We want the following to come from the *_meas catalogs as they reflect
+    # what happened in SFP calibration.
+    columnsToCopyFromMeas = ListField(dtype=str, default=["calib_",],
+                                      doc="List of \"startswith\" strings of column names to copy from "
+                                      "*_meas to *_forced_src catalog.  All columns that start with one "
+                                      "of these strings will be copied from the *_meas catalog into the "
+                                      "*_forced_src catalog.")
```
Comment (sr525) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: This confused me initially because I wasn't familiar with startswith, might be easier if it said, "list of prefixes of column names to be used with `startswith` ..", or something similar.

File: python/lsst/pipe/analysis/analysis.py [hunk 7]
```diff
@@ -118,14 +123,20 @@ def __init__(self, catalog, func, quantityName, shortName, config, qMin=-0.2, qM
...
             self.good &= np.isfinite(self.quantityError)
 
-        # Don't have flags in match and overlap catalogs (already removed in the latter)
-        if ("matches" not in self.shortName and "overlap" not in self.shortName and
-                "quiver" not in self.shortName and "inputCounts" not in self.shortName):
-            for ff in set(list(self.config.flags) + flags):
+        # Don't have flags in match and overlap catalogs (already removed in the latter).
+        # Also, if selecting calib_*_used sample, do not cull on the standard
+        # self.config.flags, only on unless explicitly set in the flags
+        # variable.
```
Comment (sr525) [inline python/lsst/pipe/analysis/analysis.py]: This doesn't quite make sense.

Comment (sr525) [inline python/lsst/pipe/analysis/analysis.py]: Not sure what the end of this comment means.

File: python/lsst/pipe/analysis/analysis.py [hunk 8]
```diff
@@ -118,14 +123,20 @@ def __init__(self, catalog, func, quantityName, shortName, config, qMin=-0.2, qM
...
+        if ("matches" not in self.shortName and "overlap" not in self.shortName and "quiver" not in
+            self.shortName and "inputCounts" not in self.shortName):
+            flagsList = flags.copy()
+            flagsList += list(self.config.flags) if self.calibUsedOnly == 0 else flagsList
```
Comment (sr525) [inline python/lsst/pipe/analysis/analysis.py]: I don't really get what this is doing. Perhaps it could be rewritten to be clearer? Does it add `self.config.flags` to `flagsList` if `self.calibUsedOnly == 0` and if not adds `flagsList` to `flagsList`?

File: python/lsst/pipe/analysis/analysis.py [hunk 9]
```diff
@@ -118,14 +123,20 @@ def __init__(self, catalog, func, quantityName, shortName, config, qMin=-0.2, qM
...
+            for ff in set(flagsList):
```
Comment (sr525) [inline python/lsst/pipe/analysis/analysis.py]: I think it would be easier to understand the code at a glance if the variable names were more informative.

File: python/lsst/pipe/analysis/analysis.py [hunk 10]
```diff
@@ -436,13 +438,13 @@ def plotAgainstMagAndHist(self, log, filename, stats=None, camera=None, ccdList=
                 axScatter.axvspan(self.magThresholdHigh, axScatter.get_xlim()[1], facecolor="k",
                                   edgecolor="none", alpha=0.10)
                 # compute running stats (just for plotting)
-                if self.calibUsedOnly == 0 and plotRunStats:
+                if plotRunStats:
                     belowThresh = data.mag < magMax  # set lower if you want to truncate plotted running stats
                     numHist, dataHist = np.histogram(data.mag[belowThresh], bins=len(xSyBins))
                     # Only plot running stats if there are a significant number of data points per bin.
-                    # Computed here as the mean number in the brightest 20% of the bins which we require
+                    # Computed here as the mean number in the brightest 10-30% of the bins which we require
```
Comment (sr525) [inline python/lsst/pipe/analysis/analysis.py]: This comment could be clearer, why the change from 20% to 10 - 30%?

File: python/lsst/pipe/analysis/plotUtils.py [hunk 11]
```diff
@@ -36,27 +36,37 @@ def __call__(self, catalog):
...
     def __init__(self, first="first_", second="second_"):
         self._first = first
         self._second = second
 
     def __call__(self, catalog1, catalog2=None):
         catalog2 = catalog2 if catalog2 else catalog1
-        first = np.where(catalog1[self._first + self._column] < 0.5, 0, 1)
-        second = np.where(catalog2[self._second + self._column] < 0.5, 0, 1)
-        return np.where(first == second, first, 2)
+        starGal1 = catalog1[self._first + self._column]
```
Comment (sr525) [inline python/lsst/pipe/analysis/plotUtils.py:63]: Some documentation or comments would be useful to understand why this is doing this and how the deliminating values were chosen.

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 1 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/coaddAnalysis.py:259]: Done.  The list needs to be created dynamically to accommodate any columns that may not exist for a given repo (certain algs not run, for example) and we don't know a-priori what filters exist for the `merge_measurement_FILTER` flags).

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 4 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: If I don't cast I get:
`TypeError: can only concatenate list (not "List") to list`
(I guess `lsst.pex.config` `ListFields` have a different effective `type`).

I defined a `needToAddColumns` variable to make this clearer.  The logic was also slightly flawed, so that got fixed too (so thanks for the request to clarify!)

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 5 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: This is in a different function and, not only are they both guaranteed to be run (e.g. if `plotMatchesOnly == True`), but I'd like to leave the possibility of allowing them to be different for the `match` catalog case, so have left as is (except to update for your preferred `columnsToCopy` predefinition as above).

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 6 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: Amended doc to omit use of `startswith`.

File: python/lsst/pipe/analysis/analysis.py (hunk 7 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/analysis.py]: Reworded to (hopefully) make sense!

Comment (laurenam) [reply python/lsst/pipe/analysis/analysis.py]: Uh...cuz it doesn't make any sense!  I've fixed it up and added more detail.

File: python/lsst/pipe/analysis/analysis.py (hunk 8 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/analysis.py]: I updated and added detail to the comment.

File: python/lsst/pipe/analysis/analysis.py (hunk 10 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/analysis.py]: Amended the comment.

File: python/lsst/pipe/analysis/plotUtils.py (hunk 11 above)
Comment (laurenam) [reply python/lsst/pipe/analysis/plotUtils.py:63]: I've added a comment (and hopefully anyone needing a bit more context will read the commit message).
